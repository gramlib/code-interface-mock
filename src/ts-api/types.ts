import ts from "typescript";

export interface TypeCacheRecord {
  kind: ts.SyntaxKind,
  aliasedTo: ts.SyntaxKind,
  node: ts.Node,
}

export type Output = Record<string | number, {}>;
export type Types = Record<string, TypeCacheRecord>;
