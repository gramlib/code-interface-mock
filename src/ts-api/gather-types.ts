import * as ts from "typescript";

interface Refs {
  [key:string]: ts.InterfaceDeclaration;
}

class Gather {
  collectTypes: Refs = {};

  gatherTypes(sourceFile: ts.SourceFile | ts.ModuleBlock) {
    const {statements} = sourceFile;
    statements.forEach(statement => {
      if (statement.kind === ts.SyntaxKind.InterfaceDeclaration) {
        const t = statement as ts.InterfaceDeclaration;
        this.collectTypes[t.name.getText()] = t;
      }
    });
  }

  clear() {
    this.collectTypes = {};
  }
}

export const gather = new Gather();
