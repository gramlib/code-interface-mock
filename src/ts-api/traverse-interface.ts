import ts from "typescript";
import {createFunctionType} from "@/mocker/mock-function-type";
import {createMockType} from "@/mocker/mock-type";
import {createMockArrayType} from "@/mocker/mock-array-type";
import {IMockType} from "@/mocker/IMockType";
import {gather} from "@/ts-api/gather-types";

export function traverseInterface(node: ts.InterfaceDeclaration) {
  const interfaceName = node.name.getText();
  const properties: Record<string, IMockType> = {};
  const {members} = node;

  // 继承
  let HeritageData = {};
  if (node.heritageClauses) {
    node.heritageClauses[0].types.forEach(it => {
      const refName = it.expression.getText();
      const interDeclare = gather.collectTypes[refName];
      if (interDeclare) {
        const { mockProperties } = traverseInterface(interDeclare);
        HeritageData = Object.assign(HeritageData, mockProperties);
      }
    });
  }

  members.forEach(it => {
    if (it.kind === ts.SyntaxKind.PropertySignature) {
      const { propertyName, propertyType } = parsePropertySignature(it as ts.PropertySignature);
      properties[propertyName] = propertyType;
    } else if (it.kind === ts.SyntaxKind.MethodSignature) {
      const { propertyName, propertyType } = parseMethodSignature(it as ts.MethodSignature);
      properties[propertyName] = propertyType;
    }
  })

  const mockProperties = {};
  Object.keys(properties).forEach(key => {
    mockProperties[key] = properties[key].mock();
  });
  const merged = Object.assign(HeritageData, mockProperties);

  return {
    interfaceName,
    mockProperties: merged
  }
}

function parsePropertySignature(node: ts.PropertySignature) {
  const propertyName = node.name.getText();
  const propertyType = processType(node.type);

  return {
    propertyName,
    propertyType
  }
}

function parseMethodSignature(node: ts.MethodSignature) {
  const propertyName = node.name.getText();
  const propertyType = processFunctionType(node);

  return {
    propertyName,
    propertyType
  }
}

function processFunctionType(node: ts.FunctionTypeNode | ts.MethodSignature) {
  const { parameters } = node;
  const params = parameters.map(it => {
    return  it.name.getText();
  });
  return createFunctionType(params);
}

function processType(type: ts.TypeNode) {
  if (type.kind === ts.SyntaxKind.FunctionType) {
    return processFunctionType(type as ts.FunctionTypeNode);
  } else if (type.kind === ts.SyntaxKind.TypeReference) {
    const t = type as ts.TypeReferenceNode;
    if (t.typeName.getText() === 'Array') {
      return createMockArrayType(t.typeArguments[0]);
    } else if (t.typeName.getText() === 'Function') {
      return createFunctionType([]);
    }else {
      return createMockType(t);
    }
  } else if (type.kind === ts.SyntaxKind.ArrayType){
    return createMockArrayType((type as ts.ArrayTypeNode).elementType);
  } else {
    return createMockType(type);
  }
}



