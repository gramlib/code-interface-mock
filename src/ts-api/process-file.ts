import ts from "typescript";
import {traverseInterface} from "@/ts-api/traverse-interface";

export function processFile(sourceFile: ts.SourceFile) {
  const collectInterfaceMock = {};

  const processNode = (node: ts.Node) => {
    if (node.kind === ts.SyntaxKind.InterfaceDeclaration) {
      const {interfaceName, mockProperties} = traverseInterface(node as ts.InterfaceDeclaration);
      collectInterfaceMock[interfaceName] = mockProperties;
    } else {
      // ignore
    }

    ts.forEachChild(node, processNode);
  };

  processNode(sourceFile);

  return collectInterfaceMock;
}
