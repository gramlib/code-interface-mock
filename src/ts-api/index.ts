import ts from "typescript";
import {gather} from "@/ts-api/gather-types";
import {processFile} from "@/ts-api/process-file";

export function mockInter(sourceFile: ts.SourceFile,) {
  gather.clear();
  gather.gatherTypes(sourceFile);

  const collectInterfaceMock = processFile(sourceFile);
  return collectInterfaceMock;
}
