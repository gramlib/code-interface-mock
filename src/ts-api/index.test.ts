import ts from "typescript";
import {mockInter} from "@/ts-api/index";

const filename = 'index.ts';
const filecontent = `\
interface Admin extends User {
   adminRecord: AdminRecord;
}

interface Student extends User {
   schoolRecord: SchoolRecord;
}

interface User {
   firstName: string;
   lastName: string;
   username: string;
   emailAddress: string;
}

interface AdminRecord {
   studentsPassedEachYear: number[];
}

interface SchoolRecord {
   startDate: string;
   endDate: string;
   isActive: boolean;
   grades: number[];
}

export interface ICommonData {
  query: Record<string, string>;
  pushWindow(url: string): void;
  alert(message: string): void;
  onBtnTap: Function;
}
`;

describe('test gather types', () => {
  it('should gather', () => {
    const sourceFile = ts.createSourceFile(filename, filecontent, ts.ScriptTarget.ES2015, true);
    const result = mockInter(sourceFile);
    expect(result).toMatchSnapshot();
  });
})
