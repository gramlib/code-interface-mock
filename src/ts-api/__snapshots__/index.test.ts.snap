// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test gather types should gather 1`] = `
Object {
  "Admin": Object {
    "adminRecord": Object {
      "studentsPassedEachYear": Array [
        96764,
        63230,
      ],
    },
    "emailAddress": "力都器史。",
    "firstName": "本她至。",
    "lastName": "织程只。",
    "username": "指会近东强受。",
  },
  "AdminRecord": Object {
    "studentsPassedEachYear": Array [
      189519,
      97719,
    ],
  },
  "ICommonData": Object {
    "alert": [Function],
    "onBtnTap": [Function],
    "pushWindow": [Function],
    "query": Object {},
  },
  "SchoolRecord": Object {
    "endDate": "出角商名。",
    "grades": Array [
      72167,
      194532,
    ],
    "isActive": true,
    "startDate": "象题色济没。",
  },
  "Student": Object {
    "emailAddress": "百根基格色想连是。",
    "firstName": "计成技却阶织向。",
    "lastName": "矿色体实大越特进。",
    "schoolRecord": Object {
      "endDate": "教权但除造土志代。",
      "grades": Array [
        10382,
        178251,
        37529,
      ],
      "isActive": true,
      "startDate": "者了过年。",
    },
    "username": "却务使天行种。",
  },
  "User": Object {
    "emailAddress": "素离及。",
    "firstName": "你据家解。",
    "lastName": "不价新员周们关之。",
    "username": "或文类标命义院。",
  },
}
`;
